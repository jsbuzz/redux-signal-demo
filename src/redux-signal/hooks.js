import { useEffect, useState } from "react";

const thunkRX = /.*=>.*dispatch.*=>/i;

const isThunk = (fn) => !!fn.toString().match(thunkRX);
const asKey = (str) => (typeof str === "function" ? str().toString() : str);

export function useActionListeners(...listeners) {
  useEffect(() => {
    listeners.reduce((a, c) => {
      if (typeof c === "function" && !isThunk(c)) {
        a.forEach((e) => {
          const key = asKey(e);
          const eventListeners = window.actionListeners[key];
          if (eventListeners) {
            eventListeners.push(c);
          } else {
            window.actionListeners[key] = [c];
          }
        });
        return [];
      }
      return [...a, c];
    }, []);

    return function cleanup() {
      console.log("before cleanup", { ...window.actionListeners });
      listeners.reduce((a, c) => {
        if (typeof c === "function" && !isThunk(c)) {
          a.forEach((e) => {
            const key = asKey(e);
            const eventListeners = window.actionListeners[key];
            eventListeners.splice(eventListeners.indexOf(c), 1);
          });
          return [];
        }
        return [...a, c];
      }, []);
      console.log("after cleanup", { ...window.actionListeners });
    };
    // eslint-disable-next-line
  }, []);
}

const asArray = (a) => (Array.isArray(a) ? a : [a]);
export function useFetchState({ fetch, done, error, errorHandler = (e) => e }) {
  const [state, setState] = useState({
    pending: false,
    error: null,
  });

  useActionListeners(
    ...asArray(fetch),
    () =>
      setState({
        pending: true,
        error: null,
      }),

    ...asArray(done),
    () =>
      setState({
        pending: false,
        error: null,
      }),

    ...asArray(error),
    (errorAction) =>
      setState({
        pending: false,
        error: errorHandler(errorAction),
      })
  );

  return [state.pending, state.error];
}

export const withPendingState = (thunk) => (...done) => {
  const [loading, setLoading] = useState(false);

  useActionListeners(...done, () => setLoading(false));

  return [
    loading,
    (...a) => {
      setLoading(true);
      return thunk(...a);
    },
  ];
};
