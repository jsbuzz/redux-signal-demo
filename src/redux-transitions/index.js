import { useEffect, useState } from "react";
import { useStore } from "react-redux";

const ACTION_LISTENERS = "_actionListeners";
const IS_THUNK = "_thunkMe";
const thunkRX = /^([^=]+|\([^)]+\))[\s]*=>[\s]*([^=]+|\([^)]+\))[\s]*=>[\s]*|^function[\s]*[^(]*\([^)]*\)[\s]*{[\s]*return[\s]+function[\s]*\([^)]*\)/i;

const thunkKey = (thunkFn) =>
  (`${thunkFn.name}`.length > 3 && thunkFn.name) || thunkFn.toString();

const isThunk = (fn) => fn[IS_THUNK] || !!fn.toString().match(thunkRX);
const asArray = (a) => (Array.isArray(a) ? a : [a]);
const actionKey = (action) =>
  typeof action === "function" ? thunkKey(action) : action.type;
const actionTypeKey = (actionType) =>
  typeof actionType === "function" ? thunkKey(actionType()) : actionType;

// you can use this property in your actions to mark them
// so the middleware will not propagate them to your reducers
export const STOP_PROPAGATION = "_stopPropagation";
export const defaultStates = {
  pending: "pending",
  success: "success",
  failure: "failure",
};

export const createActionListener = () => {
  const context = {};

  return {
    // redux middleware that will call the action listeners
    actionListener: () => (next) => (action) => {
      if (!context.store[ACTION_LISTENERS]) return next(action);

      const key = actionKey(action);
      const listeners = context.store[ACTION_LISTENERS][key];
      if (listeners) {
        listeners.forEach((listener) => listener(action));
      }

      if (!action[STOP_PROPAGATION]) next(action);
    },

    // there is no other way to access the store on the listener level
    setStore: (store) => {
      context.store = store;
    },
  };
};

const processListeners = (listeners, actionCallback) => {
  listeners.reduce((actions, current) => {
    if (typeof current !== "function" || isThunk(current)) {
      return [...actions, current];
    }
    actions.forEach((actionType) =>
      actionCallback(actionTypeKey(actionType), current)
    );
    return [];
  }, []);
};

// hook for setting listeners on actions
export function useActionListeners(...listeners) {
  const store = useStore();
  if (!store[ACTION_LISTENERS]) {
    store[ACTION_LISTENERS] = {};
  }
  useEffect(() => {
    processListeners(listeners, (event, callback) => {
      const actionListeners = store[ACTION_LISTENERS][event];
      if (actionListeners) {
        actionListeners.push(callback);
      } else {
        store[ACTION_LISTENERS][event] = [callback];
      }
    });

    return function cleanup() {
      processListeners(listeners, (event, callback) => {
        const actionListeners = store[ACTION_LISTENERS][event];
        actionListeners.splice(actionListeners.indexOf(callback), 1);
      });
    };
  }, []);
}

// hook for a standard fetch operation with pending and error states
export function usePendingState({
  pending,
  success,
  failure,
  failureHandler = (e) => e,
} = {}) {
  const [state, setState] = useState({
    pending: false,
    error: null,
  });

  useActionListeners(
    ...asArray(pending),
    () =>
      setState({
        pending: true,
        error: null,
      }),

    ...asArray(success),
    () =>
      setState({
        pending: false,
        error: null,
      }),

    ...asArray(failure),
    (errorAction) =>
      setState({
        pending: false,
        error: failureHandler(errorAction),
      })
  );

  return [state.pending, state.error];
}

// hook for transitions
export const useTransitions = (transitionStates, transitionReducer) => {
  const [state, setState] = useState(transitionReducer());

  let listeners = [];
  Object.keys(transitionStates).forEach((transition) => {
    listeners = [
      ...listeners,
      ...asArray(transitionStates[transition]),
      (action) => setState(transitionReducer(transition, action)),
    ];
  });

  useActionListeners(...listeners);

  return state;
};

// function to calculate a transition state from an action
export const mockTransition = (transitionStates, transitionReducer) => (
  action
) => {
  const transition = Object.keys(transitionStates).find(
    (transitionState) =>
      asArray(transitionStates[transitionState]).includes(action) ||
      asArray(transitionStates[transitionState]).includes(action.type)
  );

  return transitionReducer(transition, action);
};

export function thunk(thunkFn) {
  thunkFn[IS_THUNK] = true;
  thunkFn.withTransitionStates = (transitions) => {
    thunkFn.transitions =
      typeof transitions === "function" ? transitions(thunkFn) : transitions;

    return thunkFn;
  };
  return thunkFn;
}

const defaultReducer = (state, error) => [
  state === defaultStates.pending,
  state === defaultStates.failure && error,
];

export function useThunkReducer(thunkFn, reducer) {
  return useTransitions(thunkFn.transitions, reducer || defaultReducer);
}
