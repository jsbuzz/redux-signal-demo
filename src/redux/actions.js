import { STOP_PROPAGATION } from "redux-transitions";

export const FETCH_DATA_START = "FETCH_DATA_START";
export const FETCH_DATA_SUCCESS = "FETCH_DATA_SUCCESS";
export const FETCH_DATA_FAILURE = "FETCH_DATA_FAILURE";

export const fetchDataStart = () => ({
  type: FETCH_DATA_START,
  [STOP_PROPAGATION]: true,
});

export const setData = (data) => ({
  type: FETCH_DATA_SUCCESS,
  data,
});

export const fetchDataError = (error) => ({
  type: FETCH_DATA_FAILURE,
  [STOP_PROPAGATION]: true,
  error,
});

const LOGIN_SUCCESS = "LOGIN_SUCCESS";
const LOGIN_FAILURE = "LOGIN_FAILURE";

export const loginSuccess = (token) => ({
  type: LOGIN_SUCCESS,
  [STOP_PROPAGATION]: true,
  token,
});

export const loginError = (error) => ({
  type: LOGIN_FAILURE,
  [STOP_PROPAGATION]: true,
  error,
});

export const UPLOAD_CHUNK = "UPLOAD_CHUNK";
export const UPLOAD_FAILURE = "UPLOAD_FAILURE";
export const UPLOAD_SUCCESS = "UPLOAD_SUCCESS";

export const addUploadedFile = ({ fileName, fileSize }) => ({
  type: UPLOAD_SUCCESS,
  fileName,
  fileSize,
});

export const uploadError = (error) => ({
  type: UPLOAD_FAILURE,
  [STOP_PROPAGATION]: true,
  error,
});

export const uploadChunk = (percentage) => ({
  type: UPLOAD_CHUNK,
  [STOP_PROPAGATION]: true,
  percentage,
});
