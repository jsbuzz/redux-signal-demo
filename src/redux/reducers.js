import { combineReducers } from "redux";
import { FETCH_DATA_SUCCESS, UPLOAD_SUCCESS } from "./actions";

const dataReducerState = {
  data: null,
};
const dataReducer = (state = dataReducerState, action) => {
  switch (action.type) {
    case FETCH_DATA_SUCCESS:
      return {
        ...state,
        data: action.data,
      };

    default:
      console.log("dataReducer");
      return state;
  }
};

const uploadReducerState = {
  uploadedFiles: [],
  totalSize: 0,
};
const uploadReducer = (state = uploadReducerState, action) => {
  switch (action.type) {
    case UPLOAD_SUCCESS:
      return {
        ...state,
        totalSize: state.totalSize + action.fileSize,
        uploadedFiles: [
          ...state.uploadedFiles,
          {
            fileName: action.fileName,
            fileSize: action.fileSize,
          },
        ],
      };

    default:
      console.log("uploadReducer");
      return state;
  }
};

export default combineReducers({
  data: dataReducer,
  upload: uploadReducer,
});
