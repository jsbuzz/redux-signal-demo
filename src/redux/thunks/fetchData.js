import { thunk } from "../../redux-transitions";
import { setData, fetchDataError } from "../actions";

let counter = 0;
export const fetchData = thunk((delay = 800) => (dispatch) => {
  setTimeout(() => {
    if (++counter % 3) {
      dispatch(setData(`This is the fetched data #${counter}`));
    } else {
      dispatch(fetchDataError(`Whoopsie #${counter / 3}`));
    }
  }, delay);
}).withTransitionStates((fetchData) => ({
  pending: fetchData,
  success: setData().type,
  failure: fetchDataError().type,
}));
