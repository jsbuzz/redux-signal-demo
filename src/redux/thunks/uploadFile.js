import { thunk } from "../../redux-transitions";
import { addUploadedFile, uploadError, uploadChunk } from "../actions";

let counter = 1;
export const uploadFile = thunk(() => (dispatch) => {
  const step = 100 + getRandomInt(300);

  setTimeout(() => dispatch(uploadChunk(15)), step);
  setTimeout(() => dispatch(uploadChunk(35)), 2 * step);
  setTimeout(() => dispatch(uploadChunk(65)), 3 * step);
  setTimeout(() => dispatch(uploadChunk(77)), 4 * step);
  setTimeout(() => dispatch(uploadChunk(85)), 5 * step);
  setTimeout(
    () =>
      dispatch(
        step % 3
          ? addUploadedFile({
              fileName: `${counter++}_somethink.zip`,
              fileSize: (7 * step) / 100,
            })
          : uploadError(`Whoopsie #${counter}`)
      ),
    6.5 * step
  );
}).withTransitionStates((uploadFile) => ({
  pending: [uploadFile, uploadChunk().type],
  success: addUploadedFile({}).type,
  failure: uploadError().type,
}));

function getRandomInt(max) {
  return Math.floor(Math.random() * Math.floor(max));
}
