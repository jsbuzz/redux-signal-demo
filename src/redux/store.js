import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { createActionListener } from "../redux-transitions";
import reducers from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const { actionListener, setStore } = createActionListener();
const store = createStore(
  reducers,
  composeEnhancers(applyMiddleware(actionListener, thunk))
);

setStore(store);

export default store;
