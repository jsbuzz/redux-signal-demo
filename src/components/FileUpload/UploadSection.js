import React from "react";
import { useDispatch } from "react-redux";
import { uploadFile } from "../../redux/thunks/uploadFile";
import { useThunkState } from "../../redux-transitions";

export const uploadReducer = (
  state = "success",
  { error, percentage } = {}
) => ({
  isUploading: state === "pending",
  uploadError: state === "failure" && error,
  uploadPercentage: percentage || 0,
});

// this component only relies on Transition state
// @transitionState
// @dispatcher
export const UploadSection = () => {
  const dispatch = useDispatch();

  const { isUploading, uploadPercentage, uploadError } = useThunkState(
    uploadFile,
    uploadReducer
  );

  return (
    <>
      <button disabled={isUploading} onClick={() => dispatch(uploadFile())}>
        {isUploading ? "uploading..." : uploadError ? "retry" : "upload"}
      </button>
      {isUploading && (
        <progress id="file" max="100" value={uploadPercentage}>
          {uploadPercentage}% done
        </progress>
      )}
      {uploadError && (
        <span className="upload-error">
          Error while uploading: {uploadError}
        </span>
      )}
    </>
  );
};
