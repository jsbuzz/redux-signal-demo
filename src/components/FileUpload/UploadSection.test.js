import React from "react";
import { shallow } from "enzyme";
import { UploadSection, uploadReducer } from "./UploadSection";
import { mockTransition } from "redux-transitions";
import * as reduxTransitionHooks from "redux-transitions/hooks";
import * as reactRedux from "react-redux";
import { uploadChunk, uploadError, addUploadedFile } from "../../redux/actions";
import { uploadFile } from "../../redux/thunks/uploadFile";

const mockUploadTransition = mockTransition(
  uploadFile.transitions,
  uploadReducer
);

describe("uploadReducer", () => {
  it("SUCCESS_STATE and empty state should be equal", () => {
    expect(uploadReducer()).toEqual(
      uploadReducer(
        "success",
        addUploadedFile({ fileName: "test", fileSize: 99 })
      )
    );
  });
});

describe("UploadSection", () => {
  let dispatched;
  let wrapper;
  let useTransitionsStub;

  beforeEach(() => {
    dispatched = [];
    jest
      .spyOn(reactRedux, "useDispatch")
      .mockReturnValue((action) => dispatched.push(action));
    useTransitionsStub = jest.spyOn(reduxTransitionHooks, "useThunkReducer");
  });

  describe("Base transition state", () => {
    beforeEach(() => {
      useTransitionsStub.mockReturnValue({});
      wrapper = shallow(<UploadSection />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should NOT have progress bar", () => {
      expect(wrapper.find("progress").length).toEqual(0);
    });

    it("should have the CTA say 'upload'", () => {
      expect(wrapper.find("button").text()).toContain("upload");
    });

    describe("Clicking the button", () => {
      beforeEach(() => {
        wrapper.find("button").simulate("click");
      });

      it("should dispatch uploadFile", () => {
        expect(dispatched[0].toString()).toEqual(uploadFile().toString());
      });
    });
  });

  describe("Pending transition state with `uploadFile`", () => {
    beforeEach(() => {
      useTransitionsStub.mockReturnValue(mockUploadTransition(uploadFile));
      wrapper = shallow(<UploadSection />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should have a progress bar with correct value", () => {
      expect(wrapper.find("progress").prop("value")).toEqual(0);
    });

    it("should have the CTA say 'uploading'", () => {
      expect(wrapper.find("button").text()).toContain("uploading");
    });

    it("should have the CTA disabled", () => {
      expect(wrapper.find("button").prop("disabled")).toEqual(true);
    });
  });

  describe("Pending transition state with Chunk", () => {
    const chunk = 37;
    beforeEach(() => {
      useTransitionsStub.mockReturnValue(
        mockUploadTransition(uploadChunk(chunk))
      );
      wrapper = shallow(<UploadSection />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should have a progress bar with correct value", () => {
      expect(wrapper.find("progress").prop("value")).toEqual(chunk);
    });

    it("should have the CTA say 'uploading'", () => {
      expect(wrapper.find("button").text()).toContain("uploading");
    });

    it("should have the CTA disabled", () => {
      expect(wrapper.find("button").prop("disabled")).toEqual(true);
    });
  });

  describe("Error transition state", () => {
    const error = "something went wrong";

    beforeEach(() => {
      useTransitionsStub.mockReturnValue(
        mockUploadTransition(uploadError(error))
      );
      wrapper = shallow(<UploadSection />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should NOT have progress bar", () => {
      expect(wrapper.find("progress").length).toEqual(0);
    });

    it("should have the CTA say 'retry'", () => {
      expect(wrapper.find("button").text()).toContain("retry");
    });

    describe("Clicking the button", () => {
      beforeEach(() => {
        wrapper.find("button").simulate("click");
      });

      it("should dispatch uploadFile", () => {
        expect(dispatched[0].toString()).toEqual(uploadFile().toString());
      });
    });
  });
});
