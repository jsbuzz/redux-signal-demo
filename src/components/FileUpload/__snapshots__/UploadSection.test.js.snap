// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadSection Base transition state should match snapshot 1`] = `undefined`;

exports[`UploadSection Error transition state should match snapshot 1`] = `undefined`;

exports[`UploadSection Pending transition state with \`uploadFile\` should match snapshot 1`] = `undefined`;

exports[`UploadSection Pending transition state with Chunk should match snapshot 1`] = `undefined`;
